docs/pages/code/e.php

```diff
@@ -19,28 +19,25 @@
     return $value1 + $value2;
 };
 
-$fact = static function (int $number) use ($multiplication): float {
-    return Collection::range(1, $number + 1)
-        ->foldLeft(
-            $multiplication,
-            1
-        )
+$fact = static function (float $number) use ($multiplication): float {
+    return (float) Collection::range(1, $number + 1)
+        ->foldLeft($multiplication, 1)
         ->current();
```

Comment:
the `(float)` cast here should not be needed but it is because of the [return type on the interface](https://github.com/loophp/collection/blob/master/src/Contract/Operation/FoldLeftable.php#L31):
`@psalm-return \loophp\collection\Collection<TKey, T|null>`

I feel this should be tied to the return type of the callable passed to the method. Only if the callable returns `null` will the method return type be `null`, otherwise it will be `T`. However, I'm not too sure how to link the two together

Reply:
Yes, I'm sharing the same feeling. This has always been an issue: "How to properly type those reduction operations".
To be honest, I do not pretend to have yet the proper solution.

Reply:
No worries, I'll look into it some more. We might need a [type-specifying extension](https://phpstan.org/developing-extensions/type-specifying-extensions) for some of these methods. PHPStan is not running with a config file at the moment, right?

Reply:
No, nothing.

Run `./vendor/bin/grumphp run --tasks=phpstan,psalm -vvv` to see the commands that are used.

Reply:
I found this, potentially we'd want the `foldLeft` and `reduction` typed more similarly to the `reduce` here? https://arnaud.le-blanc.net/post/phpstan-generics.html

Basically saying the return type will be the type of the initial value, which should correspond to the type returned by the callback

Reply:
I see.

I started to work on a subpackage here, I guess you'll be interested to see the PSALM/PHPStan annotations: https://github.com/loophp/fpt

Reply:
Ah yes, this is the correct way: https://github.com/loophp/fpt/blob/master/src/Reduce.php#L32.

I also reached that conclusion 😄 : https://phpstan.org/r/f57b8607-c972-4d1d-bdf3-3facd22f076e

Reply:
I am reluctant to embark on significant changes of the annotations in this project until we have some "test files" for the code analysis.

What do you think of that? Basically having a dedicated folder where we just try to use the library like a user would, and ensure phpstan+psalm pass at close to the maximum levels. I'm aware we have the "examples"  docs but those are more just for showcasing the library whereas this would be part of the CI suite and errors should not be allowed

Reply:
Same tests with PSalm: https://psalm.dev/r/f07db5da95

The more tests we have, the better, so I'm all :+1: for this!